### ISSUE TITLE
Process suggestions for HED restructuring

### ISSUE THREAD

**Issue by VisLab** [hed-standard/hed-specification#30]
Here are some processes that I propose we institute to make it easier to transition retagging and validator code:

1. Each single change is in a single commit.
2. Each pull-request merge is tagged with a new version number and that we use a git annotated tag with that version number as part of that merge.
3. The person who merges a pull-request also runs the npm module auto-changelog, which automatically generates a CHANGELOG file.  (Use auto-changelog --unreleased so that it shows all the changes for each tagged version since we won't be releasing until we are done.)

I also think that we should create a **remaptag** file that consists of a two column (tab-separated) table.  This file will be automatically used to remap tags from HED 2 to HED 3.  We did something similar for the HED 1 to HED 2 transition.  A tag that needs to be changed in previously tagged data will appear in the first column and its replacement tag will be in the second column.

**Example:**
Event/ID/3      (Event, Attribute/ID/Value/3)

**Comment by bigdelys** [comment]
Agreed. 

Could you please expand on how to run the npm module? what is the exact command line? I just merged your pull request but have not run this module yet, I saw there was an auto-generated changelog file, do I need to run the module?

**Comment by VisLab** [comment]
Tags and releases are not part of the ordinary github commit process.  Care must be taken in the generation as I found out.  Further, the auto-changelog and other things assume adherence to semantic versioning.  Thus instead of vre1.0.0, I redid the tags to be like v1.0.0-restruct.
#### Merge process
Proposed update process:  
   1.  Fork the HED-restructure branch of the hed-standard/hed-specification repository to your github account.  
   2.  To avoid confusion, I have been renaming this repository hed-specification-working in my account.  
    3. Clone this repository locally to make your changes and then push to your working repository.
    4. When you are done, do a pull request for hed-specification-working to the actual repository on hed-standard.  Be sure to indicate whether this is a change that is patch level, minor, or major. Make sure that Nima, Dung, and I are reviewers of the pull request.
    5.  When all reviewers have approved, the request is merged.  

#### Adding annotated tags to define the version
After a fork has been merged, we need to create a new release so that we can control testing.  For the time being I would like to be in charge of that.  This is the process:  
    1. Clone hed-standard/hed-specification (HED-restructure) locally.
    2.  Add an annotated tag with the new version that is associated with the merge commit:: 
```
git tag -a v1.2.1-restruct   fe03asg  -m "Release v1.1.0-restruct removes IDs from non Attribute subtrees"
```
The `fe03asg` is the leading part of the hash string for the commit that we are going to associate a new version number with. Now we have to push the tag to github since the tags are stored separately from the respository:

```
git push origin --tags
```

#### Create the new release on github
Once this is done, I will go on github and create a release associated with this tag.  This is done through the GUI on github.

#### Generate the changelog:
At this point, we will generate a final change log for this release using auto-changelog.  This is an npm module.

  1.  Clone the HED-restructure branch of the hed-standard/hed-specification repository to your local machine.  
  2.  In the top level directory of the repository, execute the following:  
```
auto-changelog --commit-limit false --sort-commits date-desc

  3.  Edit the CHANGELOG.md that is created to make sure everything is in order.  
  4.  Commit locally and push to the HED-restructure branch of hed-standard/hed-specification.

**NOTE:** The only time we should be pushing to hed-standard/hed-specification is during the version/CHANGELOG process.   

Eventually we might look into continuous integration, but I don't think we are there at this point.

**Comment by dungscout96** [comment]
Github's release is often used to release an official version. Should we have releases for restructuring which is work in progress? Should we only generate release for a merge between restructuring and master instead?

**Comment by VisLab** [comment]
We need restruct releases for ongoing development of the validators.  They will be kept independent and use a different numbering scheme.

**Comment by VisLab** [comment]
The information in thread of issues was addressed by adding the information to the README.
